### ISSUE TITLE
suspicious performance regression found in #30806

### ISSUE THREAD

**Issue by hl475** [pytorch/pytorch#33334]
There is a suspicious  performance regression found in https://github.com/pytorch/pytorch/pull/30806 with `cat` operator. Creating the new issue here since we do not want to revert the PR.

To reproduce, I am using [PT OP Benchmark](https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark) with `--operators cat --tag_filter all`. environment variable `MKL_NUM_THREADS=1` and `OMP_NUM_THREADS=1`. We haven't set `KMP_AFFINITY`. For `numactl` we are using `taskset --cpu-list NUMBER` where `NUMBER` is a random int within the cpu-list range.

Here is our benchmark machine info

```
$ lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              36
On-line CPU(s) list: 0-35
Thread(s) per core:  2
Core(s) per socket:  18
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) D-2191A CPU @ 1.60GHz
Stepping:            4
CPU MHz:             1600.021
CPU max MHz:         1601.0000
CPU min MHz:         800.0000
BogoMIPS:            3200.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            25344K
NUMA node0 CPU(s):   0-35
```

For outputs from our internal run, check below. Where `before` is the current PyTorch codebase, and `after` is revert the changes in this PR.
 
```
================================================================================
Before the change, Program Output:
================================================================================
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : all

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M1_N1_K1_dim0_cpu
# Input: M: 1, N: 1, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 10.651

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M256_N512_K1_dim0_cpu
# Input: M: 256, N: 512, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 110.463

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M512_N512_K2_dim1_cpu
# Input: M: 512, N: 512, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 529.833

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim0_cpu
# Input: M: 128, N: 128, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 18.580

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim1_cpu
# Input: M: 128, N: 128, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 22.946

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim2_cpu
# Input: M: 128, N: 128, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 39.233

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim0_cpu
# Input: M: 128, N: 128, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 25.150

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim1_cpu
# Input: M: 128, N: 128, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 30.775

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim2_cpu
# Input: M: 128, N: 128, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 516.519

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim0_cpu
# Input: M: 128, N: 1024, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 111.330

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim1_cpu
# Input: M: 128, N: 1024, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 114.270

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim2_cpu
# Input: M: 128, N: 1024, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 256.177

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim0_cpu
# Input: M: 128, N: 1024, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 267.849

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim1_cpu
# Input: M: 128, N: 1024, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 269.662

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim2_cpu
# Input: M: 128, N: 1024, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 4038.911

================================================================================
After the change, Program Output:
================================================================================
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : all

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M1_N1_K1_dim0_cpu
# Input: M: 1, N: 1, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 5.776

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M256_N512_K1_dim0_cpu
# Input: M: 256, N: 512, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 75.062

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M512_N512_K2_dim1_cpu
# Input: M: 512, N: 512, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 418.434

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim0_cpu
# Input: M: 128, N: 128, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 15.653

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim1_cpu
# Input: M: 128, N: 128, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 23.845

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim2_cpu
# Input: M: 128, N: 128, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 1432.811

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim0_cpu
# Input: M: 128, N: 128, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 23.533

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim1_cpu
# Input: M: 128, N: 128, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 27.186

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim2_cpu
# Input: M: 128, N: 128, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 1440.207

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim0_cpu
# Input: M: 128, N: 1024, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 74.519

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim1_cpu
# Input: M: 128, N: 1024, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 108.705

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim2_cpu
# Input: M: 128, N: 1024, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 11414.430

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim0_cpu
# Input: M: 128, N: 1024, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 196.254

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim1_cpu
# Input: M: 128, N: 1024, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 192.988

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim2_cpu
# Input: M: 128, N: 1024, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 11472.040

================================================================================
```

_Originally posted by @hl475 in https://github.com/pytorch/pytorch/pull/30806#issuecomment-585920259_

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @ngimel

**Comment by hl475** [comment]
@mingfeima @ezyang @VitalyFedyunin @alanhdu

**Comment by mingfeima** [comment]
@hl475 I do witness performance regression for certain configuration of inputs after switching a similar CPU. On particular, the following two are most serious:
* cat_M128_N1024_K2_dim1_cpu
* cat_M512_N512_K2_dim1_cpu

The performance regression should have come from in-efficient use of L1 cache. In the **PT OP Benchmark**, the input tensor list of [cat_test.py](https://github.com/pytorch/pytorch/blob/master/benchmarks/operator_benchmark/pt/cat_test.py#L44) is actually the same tensor, e.g. [input_one, input_one].

In the old TH implementation, when iterating the length of the input tensor list, each inner line of `input_one` will be loaded only once and cached in L1 (size here is 1024 * 2 * 4 bytes).
However in #30806, the TensorIterator will copy the input tensors one after another, thus `input_one` will be loaded multiple times.

Anyway, I will provide a patch to make a fast path for such kind of scenario.

**Comment by mingfeima** [comment]
@hl475 Hi i have wrote a patch to speedup on particular condition on CPU path, could you please help verify if the [patch](https://github.com/mingfeima/pytorch/commit/fe9dcc47fd857ddfad25c764e2b571e7ab7d80bc) works on your machine.

Basically the code will fall into this path in case:
* single thread run or the `numel` is smaller that `grain_size`
* all `inputs` are contiguous
* `inputs` are float or double

Here is the performance compare on my machine,
```bash
(pytorch-mingfei) [mingfei@caffe2_mlt temp]$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                72
On-line CPU(s) list:   0-71
Thread(s) per core:    2
Core(s) per socket:    18
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 5220 CPU @ 2.20GHz
Stepping:              6
CPU MHz:               1000.097
CPU max MHz:           2200.0000
CPU min MHz:           1000.0000
BogoMIPS:              4400.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-17,36-53
NUMA node1 CPU(s):     18-35,54-71
```

**reverted** - after reverting #30806
**pr30806** - current pytorch master
**fixed** - with the fix patch
**diff** - performance speedup of **fixed** over **reverted**

Test   Name | reverted | pr30806 | fixed | speedup
-- | -- | -- | -- | --
cat_M1_N1_K1_dim0_cpu | 5.593 | 9.198 | 7.207 | 0.78
cat_M256_N512_K1_dim0_cpu | 127.33 | 84.292 | 80.625 | 1.58
cat_M512_N512_K2_dim1_cpu | 353.054 | 413.959 | 301.693 | 1.17
cat_M128_N128_K1_dim0_cpu | 12.68 | 14.221 | 11.746 | 1.08
cat_M128_N128_K1_dim1_cpu | 18.399 | 16.733 | 11.841 | 1.55
cat_M128_N128_K1_dim2_cpu | 1005.457 | 32.53 | 97.967 | 10.26
cat_M128_N128_K2_dim0_cpu | 17.653 | 20.597 | 15.988 | 1.10
cat_M128_N128_K2_dim1_cpu | 24.994 | 22.352 | 15.68 | 1.59
cat_M128_N128_K2_dim2_cpu | 1014.369 | 412.458 | 112.919 | 8.98
cat_M128_N1024_K1_dim0_cpu | 139.737 | 93.038 | 78.427 | 1.78
cat_M128_N1024_K1_dim1_cpu | 90.152 | 89.906 | 79.991 | 1.13
cat_M128_N1024_K1_dim2_cpu | 8054.711 | 198.747 | 732.695 | 10.99
cat_M128_N1024_K2_dim0_cpu | 310.763 | 238.977 | 202.016 | 1.54
cat_M128_N1024_K2_dim1_cpu | 162.64 | 203.8 | 151.585 | 1.07
cat_M128_N1024_K2_dim2_cpu | 8143.899 | 3226.182 | 854.913 | 9.53

Feel free to drop any comment.

**Comment by hl475** [comment]
Thanks @mingfeima. I will use your patch and run it today. Will keep you posted.

**Comment by hl475** [comment]
Thanks for your contribution to PyTorch @mingfeima! I patched the PR (with @houseroad's help https://github.com/pytorch/pytorch/pull/33499) and ran it on our internal machine. The configs of the machine are same as before, the numbers shows performance improvement for most of the cases besides two cases (cat_M128_N1024_K1_dim2 392.77% worse and cat_M128_N128_K1_dim2 330.02% worse). With the changes in this patch and  #30806, I think it look good to me :)

Here are the perf numbers (before is today's pytorch codebase and after is with patching your diff):
```
================================================================================
Before the change, Program Output:
================================================================================
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : all

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M1_N1_K1_dim0_cpu
# Input: M: 1, N: 1, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 10.251

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M256_N512_K1_dim0_cpu
# Input: M: 256, N: 512, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 110.233

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M512_N512_K2_dim1_cpu
# Input: M: 512, N: 512, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 526.581

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim0_cpu
# Input: M: 128, N: 128, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 17.545

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim1_cpu
# Input: M: 128, N: 128, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 21.872

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim2_cpu
# Input: M: 128, N: 128, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 37.957

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim0_cpu
# Input: M: 128, N: 128, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 24.547

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim1_cpu
# Input: M: 128, N: 128, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 30.086

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim2_cpu
# Input: M: 128, N: 128, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 516.245

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim0_cpu
# Input: M: 128, N: 1024, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 107.419

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim1_cpu
# Input: M: 128, N: 1024, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 112.745

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim2_cpu
# Input: M: 128, N: 1024, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 255.153

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim0_cpu
# Input: M: 128, N: 1024, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 267.247

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim1_cpu
# Input: M: 128, N: 1024, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 269.323

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim2_cpu
# Input: M: 128, N: 1024, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 4039.966

================================================================================
After the change, Program Output:
================================================================================
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : all

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M1_N1_K1_dim0_cpu
# Input: M: 1, N: 1, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 7.496

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M256_N512_K1_dim0_cpu
# Input: M: 256, N: 512, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 106.783

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M512_N512_K2_dim1_cpu
# Input: M: 512, N: 512, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 410.802

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim0_cpu
# Input: M: 128, N: 128, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 15.053

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim1_cpu
# Input: M: 128, N: 128, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 15.521

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K1_dim2_cpu
# Input: M: 128, N: 128, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 165.406

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim0_cpu
# Input: M: 128, N: 128, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 21.601

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim1_cpu
# Input: M: 128, N: 128, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 21.525

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N128_K2_dim2_cpu
# Input: M: 128, N: 128, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 186.174

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim0_cpu
# Input: M: 128, N: 1024, K: 1, dim: 0, device: cpu
Forward Execution Time (us) : 107.576

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim1_cpu
# Input: M: 128, N: 1024, K: 1, dim: 1, device: cpu
Forward Execution Time (us) : 106.557

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K1_dim2_cpu
# Input: M: 128, N: 1024, K: 1, dim: 2, device: cpu
Forward Execution Time (us) : 1261.451

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim0_cpu
# Input: M: 128, N: 1024, K: 2, dim: 0, device: cpu
Forward Execution Time (us) : 268.425

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim1_cpu
# Input: M: 128, N: 1024, K: 2, dim: 1, device: cpu
Forward Execution Time (us) : 208.170

# Benchmarking PyTorch: cat
# Mode: Eager
# Name: cat_M128_N1024_K2_dim2_cpu
# Input: M: 128, N: 1024, K: 2, dim: 2, device: cpu
Forward Execution Time (us) : 1427.657

================================================================================
```

**Comment by peterbell10** [comment]
It looks like #33534 has fixed this and the benchmarks look better on my machine, so closing.
